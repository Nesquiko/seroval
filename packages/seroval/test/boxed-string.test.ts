import { describe, it, expect } from 'vitest';
import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromJSON,
  serialize,
  serializeAsync,
  toJSON,
  toJSONAsync,
} from '../src';

describe('boxed string', () => {
  describe('serialize', () => {
    it('supports boxed strings', () => {
      expect(serialize(Object('"hello"'))).toMatchSnapshot();
      expect(serialize(Object('<script></script>'))).toMatchSnapshot();
      expect(deserialize<object>(serialize(Object('"hello"'))).valueOf()).toBe('"hello"');
      expect(deserialize<object>(serialize(Object('<script></script>'))).valueOf()).toBe('<script></script>');
    });
  });
  describe('serializeAsync', () => {
    it('supports boxed strings', async () => {
      expect(await serializeAsync(Promise.resolve(Object('"hello"')))).toMatchSnapshot();
      expect(await serializeAsync(Promise.resolve(Object('<script></script>')))).toMatchSnapshot();
    });
  });
  describe('toJSON', () => {
    it('supports boxed strings', () => {
      expect(JSON.stringify(toJSON(Object('"hello"')))).toMatchSnapshot();
      expect(JSON.stringify(toJSON(Object('<script></script>')))).toMatchSnapshot();
      expect(fromJSON<object>(toJSON(Object('"hello"'))).valueOf()).toBe('"hello"');
      expect(fromJSON<object>(toJSON(Object('<script></script>'))).valueOf()).toBe('<script></script>');
    });
  });
  describe('toJSONAsync', () => {
    it('supports boxed strings', async () => {
      expect(
        JSON.stringify(await toJSONAsync(Promise.resolve(Object('"hello"')))),
      ).toMatchSnapshot();
      expect(
        JSON.stringify(await toJSONAsync(Promise.resolve(Object('<script></script>')))),
      ).toMatchSnapshot();
    });
  });
  describe('crossSerialize', () => {
    it('supports boxed strings', () => {
      expect(crossSerialize(Object('"hello"'))).toMatchSnapshot();
      expect(crossSerialize(Object('<script></script>'))).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports boxed strings', () => {
        expect(crossSerialize(Object('"hello"'), { scopeId: 'example' })).toMatchSnapshot();
        expect(crossSerialize(Object('<script></script>'), { scopeId: 'example' })).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports boxed strings', async () => {
      expect(await crossSerializeAsync(Promise.resolve(Object('"hello"')))).toMatchSnapshot();
      expect(await crossSerializeAsync(Promise.resolve(Object('<script></script>')))).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports boxed strings', async () => {
        expect(await crossSerializeAsync(Promise.resolve(Object('"hello"')), { scopeId: 'example' })).toMatchSnapshot();
        expect(await crossSerializeAsync(Promise.resolve(Object('<script></script>')), { scopeId: 'example' })).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports boxed strings', async () => new Promise<void>((done) => {
      crossSerializeStream(Promise.resolve(Object('"hello"')), {
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          done();
        },
      });
    }));
    it('supports boxed sanitized strings', async () => new Promise<void>((done) => {
      crossSerializeStream(Promise.resolve(Object('<script></script>')), {
        onSerialize(data) {
          expect(data).toMatchSnapshot();
        },
        onDone() {
          done();
        },
      });
    }));
    describe('scoped', () => {
      it('supports boxed strings', async () => new Promise<void>((done) => {
        crossSerializeStream(Promise.resolve(Object('"hello"')), {
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            done();
          },
        });
      }));
      it('supports boxed sanitized strings', async () => new Promise<void>((done) => {
        crossSerializeStream(Promise.resolve(Object('<script></script>')), {
          scopeId: 'example',
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            done();
          },
        });
      }));
    });
  });
});
