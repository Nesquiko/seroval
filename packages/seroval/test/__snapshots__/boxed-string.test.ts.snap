// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`boxed string > crossSerialize > scoped > supports boxed strings 1`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=Object(\\"\\\\\\"hello\\\\\\"\\"))($R[\\"example\\"]))"`;

exports[`boxed string > crossSerialize > scoped > supports boxed strings 2`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\"))($R[\\"example\\"]))"`;

exports[`boxed string > crossSerialize > supports boxed strings 1`] = `"(self.$R=self.$R||[],$R[0]=Object(\\"\\\\\\"hello\\\\\\"\\"))"`;

exports[`boxed string > crossSerialize > supports boxed strings 2`] = `"(self.$R=self.$R||[],$R[0]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\"))"`;

exports[`boxed string > crossSerializeAsync > scoped > supports boxed strings 1`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=Promise.resolve($R[1]=Object(\\"\\\\\\"hello\\\\\\"\\")))($R[\\"example\\"]))"`;

exports[`boxed string > crossSerializeAsync > scoped > supports boxed strings 2`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=Promise.resolve($R[1]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\")))($R[\\"example\\"]))"`;

exports[`boxed string > crossSerializeAsync > supports boxed strings 1`] = `"(self.$R=self.$R||[],$R[0]=Promise.resolve($R[1]=Object(\\"\\\\\\"hello\\\\\\"\\")))"`;

exports[`boxed string > crossSerializeAsync > supports boxed strings 2`] = `"(self.$R=self.$R||[],$R[0]=Promise.resolve($R[1]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\")))"`;

exports[`boxed string > crossSerializeStream > scoped > supports boxed sanitized strings 1`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R[\\"example\\"]))"`;

exports[`boxed string > crossSerializeStream > scoped > supports boxed sanitized strings 2`] = `"($R=>($R[3]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\")))($R[\\"example\\"])"`;

exports[`boxed string > crossSerializeStream > scoped > supports boxed strings 1`] = `"((self.$R=self.$R||{})[\\"example\\"]=[],($R=>$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())($R[\\"example\\"]))"`;

exports[`boxed string > crossSerializeStream > scoped > supports boxed strings 2`] = `"($R=>($R[3]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=Object(\\"\\\\\\"hello\\\\\\"\\")))($R[\\"example\\"])"`;

exports[`boxed string > crossSerializeStream > supports boxed sanitized strings 1`] = `"(self.$R=self.$R||[],$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())"`;

exports[`boxed string > crossSerializeStream > supports boxed sanitized strings 2`] = `"($R[3]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=Object(\\"\\\\x3Cscript>\\\\x3C/script>\\"))"`;

exports[`boxed string > crossSerializeStream > supports boxed strings 1`] = `"(self.$R=self.$R||[],$R[0]=($R[1]=(s,f,p)=>((p=new Promise((a,b)=>{s=a,f=b})).s=s,p.f=f,p))())"`;

exports[`boxed string > crossSerializeStream > supports boxed strings 2`] = `"($R[3]=(p,d)=>{p.s(d),p.status=\\"success\\",p.value=d;delete p.s;delete p.f})($R[0],$R[2]=Object(\\"\\\\\\"hello\\\\\\"\\"))"`;

exports[`boxed string > serialize > supports boxed strings 1`] = `"Object(\\"\\\\\\"hello\\\\\\"\\")"`;

exports[`boxed string > serialize > supports boxed strings 2`] = `"Object(\\"\\\\x3Cscript>\\\\x3C/script>\\")"`;

exports[`boxed string > serializeAsync > supports boxed strings 1`] = `"Promise.resolve(Object(\\"\\\\\\"hello\\\\\\"\\"))"`;

exports[`boxed string > serializeAsync > supports boxed strings 2`] = `"Promise.resolve(Object(\\"\\\\x3Cscript>\\\\x3C/script>\\"))"`;

exports[`boxed string > toCrossJSON > supports boxed strings 1`] = `"{\\"t\\":27,\\"i\\":0,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\\\\\\\"hello\\\\\\\\\\\\\\"\\"}}"`;

exports[`boxed string > toCrossJSON > supports boxed strings 2`] = `"{\\"t\\":27,\\"i\\":0,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\x3Cscript>\\\\\\\\x3C/script>\\"}}"`;

exports[`boxed string > toCrossJSONAsync > supports boxed strings 1`] = `"{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":27,\\"i\\":1,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\\\\\\\"hello\\\\\\\\\\\\\\"\\"}}}"`;

exports[`boxed string > toCrossJSONAsync > supports boxed strings 2`] = `"{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":27,\\"i\\":1,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\x3Cscript>\\\\\\\\x3C/script>\\"}}}"`;

exports[`boxed string > toCrossJSONStream > supports boxed sanitized strings 1`] = `"{\\"t\\":28,\\"i\\":0,\\"f\\":{\\"t\\":41,\\"i\\":1,\\"s\\":2}}"`;

exports[`boxed string > toCrossJSONStream > supports boxed sanitized strings 2`] = `"{\\"t\\":29,\\"i\\":0,\\"a\\":[{\\"t\\":41,\\"i\\":3,\\"s\\":3},{\\"t\\":27,\\"i\\":2,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\x3Cscript>\\\\\\\\x3C/script>\\"}}]}"`;

exports[`boxed string > toCrossJSONStream > supports boxed strings 1`] = `"{\\"t\\":28,\\"i\\":0,\\"f\\":{\\"t\\":41,\\"i\\":1,\\"s\\":2}}"`;

exports[`boxed string > toCrossJSONStream > supports boxed strings 2`] = `"{\\"t\\":29,\\"i\\":0,\\"a\\":[{\\"t\\":41,\\"i\\":3,\\"s\\":3},{\\"t\\":27,\\"i\\":2,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\\\\\\\"hello\\\\\\\\\\\\\\"\\"}}]}"`;

exports[`boxed string > toJSON > supports boxed strings 1`] = `"{\\"t\\":{\\"t\\":27,\\"i\\":0,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\\\\\\\"hello\\\\\\\\\\\\\\"\\"}},\\"f\\":16383,\\"m\\":[]}"`;

exports[`boxed string > toJSON > supports boxed strings 2`] = `"{\\"t\\":{\\"t\\":27,\\"i\\":0,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\x3Cscript>\\\\\\\\x3C/script>\\"}},\\"f\\":16383,\\"m\\":[]}"`;

exports[`boxed string > toJSONAsync > supports boxed strings 1`] = `"{\\"t\\":{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":27,\\"i\\":1,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\\\\\\\"hello\\\\\\\\\\\\\\"\\"}}},\\"f\\":16383,\\"m\\":[]}"`;

exports[`boxed string > toJSONAsync > supports boxed strings 2`] = `"{\\"t\\":{\\"t\\":12,\\"i\\":0,\\"s\\":1,\\"f\\":{\\"t\\":27,\\"i\\":1,\\"f\\":{\\"t\\":1,\\"s\\":\\"\\\\\\\\x3Cscript>\\\\\\\\x3C/script>\\"}}},\\"f\\":16383,\\"m\\":[]}"`;
